src/react-imgix.jsx

```diff
@@ -109,68 +110,97 @@ function buildSrc({
-
-      let srcFn = buildDprSrcWithQualityByDpr;
-      if (q) {
-        srcFn = buildDprSrcWithQuality;
-      } else if (disableQualityByDPR) {
-        srcFn = buildDprSrcWithoutQuality;
+    for (const k in urlParams) {
+      if (PARAM_EXPANSION[k]) {
+        params[PARAM_EXPANSION[k]] = urlParams[k];
+      } else {
+        params[k] = urlParams[k];
       }
+    }
 
-      srcSet = "";
-      const len = DPR_QUALITY_VALUES.length;
-      for (let i = 0; i < len; i++) {
-        const quality = DPR_QUALITY_VALUES[i];
-        srcSet += srcFn(constructedUrl, q || quality, i + 1) + ", ";
-      }
-      srcSet = srcSet.slice(0, -2);
-    } else {
-      const { width, w, height, ...urlParams } = srcOptions;
-      const constructedUrl = constructUrl(rawSrc, urlParams);
+    if (width) {
+      params["w"] = width;
+    }
 
-      const aspectRatio = imgixParams.ar;
-      let showARWarning =
-        aspectRatio != null && aspectRatioIsValid(aspectRatio) === false;
+    if (height) {
+      params["h"] = height;
+    }
 
-      let srcFn = buildSrcSetPairWithTargetWidth;
-      if (height) {
-        srcFn = buildSrcSetPairWithFixedHeight;
-      }
+    const [scheme, rest] = rawSrc.split("://");
+    const [domain, ...pathComponents] = rest.split("/");
+    let useHTTPS = scheme == "https";
+
+    const client = new ImgixClient({
+      domain: domain,
+      useHTTPS: useHTTPS,
+      includeLibraryParam: false,
+    });
 
-      srcSet = "";
-      const len = targetWidths.length;
-      for (let i = 0; i < len; i++) {
-        const targetWidth = targetWidths[i];
-        srcSet += srcFn(constructedUrl, targetWidth, height) + ", ";
+    const srcSet = client.buildSrcSet(
+      pathComponents.join("/"),
+      { ...params, q },
+      {
+        disableVariableQuality: disableQualityByDPR,
       }
-      srcSet = srcSet.slice(0, -2);
+    );
 
-      if (
-        NODE_ENV !== "production" &&
-        showARWarning &&
-        config.warnings.invalidARFormat
-      ) {
-        console.warn(
-          `[Imgix] The aspect ratio passed ("${aspectRatio}") is not in the correct format. The correct format is "W:H".`
-        );
+    const src = constructUrl(rawSrc, srcOptions);
+    return { src, srcSet };
+  } else {
+    const { width, w, height, h, ...urlParams } = srcOptions;
+    const params = {};
+
+    for (const k in urlParams) {
+      if (PARAM_EXPANSION[k]) {
+        params[PARAM_EXPANSION[k]] = urlParams[k];
+      } else {
+        params[k] = urlParams[k];
       }
     }
-  }
 
-  return {
-    src,
-    srcSet,
-  };
+    if (width) {
+      params["w"] = width ? width : w;
+    }
+
+    if (height) {
+      params["h"] = height || h;
+    }
+
+    const [scheme, rest] = rawSrc.split("://");
+    const [domain, ...pathComponents] = rest.split("/");
+    let useHTTPS = scheme == "https";
+
+    const client = new ImgixClient({
+      domain: domain,
+      useHTTPS: useHTTPS,
+      includeLibraryParam: false,
+    });
```

Comment:
I noticed `ImgixClient` gets setup several times throughout this component. I'm honestly not sure if it has any bearing on performance, but I get the feeling it should be pulled up more towards the top level of the component to avoid repeated work.

Reply:
I don't think it impacts performance because we only setup a client once either way (once in each block of the if-else blocks). The if-else blocks are kind of long, so I made each do one job (keeping things scoped/encapsulated into each block) then each block returns it's result. In general, I try no to use a single, top-level definition, through multiple long blocks of code (this is defensive, i.e. to prevent reassignment or some mutation that might make debugging difficult later on).

That said, I think we can shorten things up a lot here. Now that this initial bit has been looked over I'll make another pass over things refactoring/cleaning up.